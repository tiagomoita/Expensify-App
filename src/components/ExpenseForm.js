import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

//const now = moment();
//console.log(now.format('MMMM Do, YYYY'));

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };

    }

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount.' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    onDateChange = (date) => {
        if(date) {
            this.setState(() => ({ createdAt: date }))
        }
    };

    onFocusChange = ({focused}) => {
        this.setState(() => ({ calenderFocused: focused }));
    };


    render() {
        console.log(this.state.calenderFocused);
        return (
            <div>
                {this.state.error}
                <form onSubmit={this.onSubmit}>
                    <input
                        type='text'
                        value={this.state.description}
                        placeholder='Description'
                        autoFocus
                        onChange={(newText) => {
                            const description = newText.target.value;
                            this.setState(() => ({ description }));
                        }}
                    />
                    <input
                        type='text'
                        value={this.state.amount}
                        placeholder='Amount'
                        onChange={(newAmount) => {
                            if (!newAmount.target.value || newAmount.target.value.match(/^\d{1,}(\.\d{0,2})?$/)) {
                                const amount = newAmount.target.value;
                                this.setState(() => ({ amount }));
                            }
                        }}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calenderFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder='Add a note for your expense(optional)'
                        value={this.state.note}
                        onChange={(newNote) => { this.setState(() => ({ note: newNote.target.value })) }}
                    >
                    </textarea>
                    {this.props.expense ? (<button>Save Expense</button>) : <button>Add Expense</button>}
                </form>
            </div >
        );
    }
};

export default ExpenseForm;
