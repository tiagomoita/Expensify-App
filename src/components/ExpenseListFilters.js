import React, { useState } from 'react';
import * as actions from './../redux/actions';
import { connect, useDispatch } from 'react-redux';
import { DateRangePicker } from 'react-dates';

const ExpenseListFilters = props => {
    const { text, sortBy, startDate, endDate } = props.filters;
    const dispatch = useDispatch();
    const [calendarFocused, setCalendarFocused] = useState(null);

    return (
        <div>
            <input type='text' value={text} onChange={(newText) => {
                dispatch(actions.setTextFilter(newText.target.value));
            }} />
            <select value={sortBy} onChange={(newSort) => {
                if (newSort.target.value === 'date') {
                    dispatch(actions.sortByDate());
                }
                else if (newSort.target.value === 'amount') {
                    dispatch(actions.sortByAmount());
                }
            }}>
                <option value='date'>Date</option>
                <option value='amount'>Amount</option>
            </select>
            <DateRangePicker
                startDateId="start"
                endDateId="end"
                startDate={startDate}
                endDate={endDate}
                onDatesChange={({ startDate, endDate}) => {
                    dispatch(actions.setStartDate(startDate))
                    dispatch(actions.setEndDate(endDate))
                }}
                focusedInput={calendarFocused}
                onFocusChange={(focusedInput) => {
                    setCalendarFocused(focusedInput)
                }}
                numberOfMonths={1}
                isOutsideRange={() => false}
                showClearDates={true}
            />
        </div>
    );
};

const mapStateToProps = state => {
    return {
        filters: state.filters
    };
};

export default connect(mapStateToProps, actions)(ExpenseListFilters);
