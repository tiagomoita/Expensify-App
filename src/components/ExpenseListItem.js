import React from 'react';
import { Link } from 'react-router-dom';

export const ExpenseListItem = props => {
    const { id, description, amount, createdAt } = props;

    return (
        <div>
            <Link to={`/edit/${id}`}>
                <h1>{description}</h1>
            </Link>
            <p>amount: {amount}â‚¬ - createdAt: {createdAt}</p>
        </div>
    );
};


export default ExpenseListItem;
