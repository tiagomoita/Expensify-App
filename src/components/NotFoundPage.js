import React from 'react';
import { Link } from 'react-router-dom';

//Link -> Client-Side Routing
export const NotFoundPage = () => {
    return (
        <div>
            404 - <Link to="/">Go Home</Link>
        </div>
    );
};

export default NotFoundPage;