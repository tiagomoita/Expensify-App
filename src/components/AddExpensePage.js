import React from 'react';
import ExpenseForm from './ExpenseForm';
import * as actions from './../redux/actions';
import { connect } from 'react-redux';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        //const { description, note, amount, createdAt } = expense;
        this.props.addExpense(expense);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        );
    }
};

function mapDispatchToProps(dispatch) {
    return {
        addExpense: (expense) => dispatch(actions.addExpense(expense)),
        //dispatch,
        //actions
    };
}


export default connect(null, mapDispatchToProps)(AddExpensePage);