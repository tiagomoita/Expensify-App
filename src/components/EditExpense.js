import React from 'react';
import ExpenseForm from './ExpenseForm';
import * as actions from './../redux/actions';
import { connect, useDispatch } from 'react-redux';

const EditExpense = props => {
    const { id } = props.match.params;
    const dispatch = useDispatch();

    return (
        <div>
            <ExpenseForm
                expense={props.expense}
                onSubmit={(expense) => {
                    dispatch(actions.editExpense(id, expense));
                    props.history.push('/');
                }}
            />
            <button onClick={() => {
                dispatch(actions.removeExpense({ id }));
                props.history.push('/');
            }}>Remove</button>
        </div>
    );
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};


export default connect(mapStateToProps, actions)(EditExpense);