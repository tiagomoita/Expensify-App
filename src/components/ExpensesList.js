import React from 'react';
import * as actions from './../redux/actions';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from './../redux/selectors/expenses';

export const ExpensesList = props => {


    const { expenses } = props;
    return (
        <div>
            { expenses.length === 0 ? (
                <p>No Expenses</p>
            ) : (
                expenses.map((expense, index) => {
                    return (
                        //NOT A GOOD IDEA TO USE INDEX AS A KEY, IS BETTER TO USE THE ID
                        <ExpenseListItem key={index} {...expense} />
                    );
                })
            )}
        </div>
    );
};

const mapStateToProps = state => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps, actions)(ExpensesList);