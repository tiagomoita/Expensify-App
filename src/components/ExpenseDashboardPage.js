import React from 'react';
import ExpensesList from './ExpensesList';
import ExpenseListFilters from './ExpenseListFilters';

export const ExpenseDashboardPage = () => {
    return (
        <div>
            <ExpenseListFilters />
            <ExpensesList />
        </div>
    );
};

export default ExpenseDashboardPage;