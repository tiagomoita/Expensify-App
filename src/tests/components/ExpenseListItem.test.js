import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseListItem } from '../../components/ExpenseListItem';
import expenses from '../redux/fixtures/expenses';


test('should render ExpenseListItem with expense', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>)
    expect(wrapper).toMatchSnapshot();
});