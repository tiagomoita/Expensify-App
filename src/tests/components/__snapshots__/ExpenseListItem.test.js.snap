// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expense 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h1>
      Gum
    </h1>
  </Link>
  <p>
    amount: 
    195
    â‚¬ - createdAt: 
    0
  </p>
</div>
`;
